File: Arrays/CountSpecialTriplets.cpp
class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        int MOD = 1e9 + 7;
        int MAX_VAL = 100000;
        int n = nums.size();

        vector<long long> prefix(MAX_VAL + 1, 0);
        vector<long long> suffix(MAX_VAL + 1, 0);

        long long total = 0;

        for (int num: nums){
            suffix[num]++;
        }

        for(int j = 0; j < n; j++){
            int val = nums[j];
            int target = 2 * val;

            suffix[val]--;

            if(target <= MAX_VAL){
                total = (total + prefix[target] * suffix[target]) % MOD;
            }

            prefix[val]++;
        }
        return total;
    }
};

File: Arrays/CountSpecialTriplets.py
class Solution:
    def specialTriplets(self, nums: List[int]) -> int:
        MOD = 10**9 + 7
        n = len(nums)
        
        suffix_cnt = defaultdict(int) 
        for num in nums:
            suffix_cnt[num] += 1
        
        prefix_cnt = defaultdict(int)

        total_triplets = 0

        for j in range(n):
            val = nums[j]
            target = 2 * val

            suffix_cnt[val] -= 1

            total_triplets += prefix_cnt[target] * suffix_cnt[target]
            total_triplets %= MOD

            prefix_cnt[val] += 1

        return total_triplets % MOD